drivers/adc/ad400x/iio_ad400x.c

```diff
@@ -0,0 +1,236 @@
+#ifdef IIO_SUPPORT
+
+#include <stdio.h>
+#include <math.h>
+#include <string.h>
+
+#include "ad400x.h"
+#include "iio_ad400x.h"
+#include "iio.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+static struct scan_type ad400x_iio_scan_types[] = {
+	[ID_AD4000] = { .sign = 'u', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4001] = { .sign = 's', .realbits = 16, .storagebits = 32, .is_big_endian = true, .shift = 16},
+	[ID_AD4002] = { .sign = 'u', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4003] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4004] = { .sign = 'u', .realbits = 16, .storagebits = 32, .is_big_endian = true, .shift = 16},
+	[ID_AD4005] = { .sign = 's', .realbits = 16, .storagebits = 32, .is_big_endian = true, .shift = 16},
+	[ID_AD4006] = { .sign = 'u', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4007] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4011] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4020] = { .sign = 's', .realbits = 20, .storagebits = 32, .is_big_endian = true, .shift = 12},
+	[ID_ADAQ4003] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+};
+
+/**
+ * @brief Read the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to read data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad400x_debug_reg_read(void *device, uint32_t reg,
+		uint32_t *val)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	int ret;
+	uint8_t val8;
+
+	ret = ad400x_spi_reg_read(dev, &val8);
+	if (ret)
+		return ret;
+
+	*val = val8;
+	return 0;
+}
+
+/**
+ * @brief Write the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to write data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad400x_debug_reg_write(void *device, uint32_t reg,
+		uint32_t val)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+
+	return ad400x_spi_reg_write(dev, (uint8_t)val);
+}
+
+static int get_raw(void *device, char *buf, uint32_t len,
+		   const struct iio_ch_info *channel, intptr_t id)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	struct scan_type *scan_type = &ad400x_iio_scan_types[dev->dev_id];
+	uint32_t results;
+	int ret;
+
+	ret = ad400x_spi_single_conversion(dev, &results);
+	if (ret)
+		return ret;
+
+	no_os_memswap64(&results, 4, 4);
+	results >>= scan_type->shift;
```

Comment:
based on the raw data type format (2's complement or binary), you will also have to get the sign and append the result with the sign bit. 
this is correct for binary data, for 2's complement you will have to use sign_extend (see examples in other drivers)